### PR TITLE
Use hashids for smaller keys and less memory usage, fix #288

### LINKED ISSUE #288: Reduce the size of ids to reduce memory usage
I'm surprised we don't have an issue for that but we already made some tries.

At Confoo, I was attending [a session about Redis](https://confoo.ca/en/yul2017/session/get-more-than-a-cache-back-redis) ([slides](https://www.slideshare.net/maartenba/get-more-than-a-cache-back-confoo-montreal)) and Maarten said the use of `incr` can be used to create distributed unique ids. Might be interesting with our asynchronous loading to generate an id from the position.

### PR DESCRIPTION
**Pull request by davidbgk**
Loading `BAN_odbl_76-json`:

Before (master): `Done: 35023 | Elapsed: 0:01:16 | 1m 20s 349ms`

with:

```
> DBINFO
keyspace_misses: 62433
keyspace_hits: 4453
used_memory_human: 297.95M
total_commands_processed: 9859055
total_connections_received: 83
connected_clients: 1
nb keys (db 0): 108358
nb keys (db 1): 35023
```

After: `Done: 35023 | Elapsed: 0:01:08 | 1m 12s 522ms`

with:

```
> DBINFO
keyspace_misses: 62433
keyspace_hits: 4453
used_memory_human: 279.78M
total_commands_processed: 5006893
total_connections_received: 66
connected_clients: 1
nb keys (db 0): 108359
nb keys (db 1): 35023
```

The benchmark is not really relevant from one run to another but we don't have a significative loss. The memory has reduced 🎉 and the difference will be more important with more keys!

I wonder if the `_` (underscore) to prefix `hashid` is pertinent.

### CODE REVIEW

File: addok/core.py [hunk 1]
```diff
@@ -102,9 +102,9 @@ def str_distance(self):
         return self._scores.get('str_distance', [0.0])[0]
 
     @classmethod
-    def from_id(self, _id):
-        """Return a result from it's document id."""
-        return Result(keys.document_key(_id))
+    def from_hashid(self, hashid):
```
Comment (davidbgk) [inline addok/core.py]: I don't know if we want to keep `from_(hash)id`, especially if we remove https://github.com/addok/addok/issues/254 soon.

File: addok/db.py [hunk 2]
```diff
@@ -13,6 +17,10 @@ def connect(self, *args, **kwargs):
     def __getattr__(self, name):
         return getattr(self.instance, name)
 
+    def next_hashid(self):
+        _counter = self.incr('_counter')
```
Comment (davidbgk) [inline addok/db.py]: Do we have/set a pattern for internal keys? Like `i|counter`?

File: addok/ds.py [hunk 3]
```diff
@@ -56,7 +56,9 @@ def store_documents(docs):
     for doc in docs:
         if not doc:
             continue
-        key = keys.document_key(doc['id'])
+        if '_hashid' not in doc:
```
Comment (davidbgk) [inline addok/ds.py]: I don't know if we want to use the `_` (underscore) prefix here and if we want the end-user to know about it. Related to the GET possibility.

File: addok/shell.py [hunk 4]
```diff
@@ -178,9 +178,10 @@ def format_scores(result):
                 return result.score
 
         for result in results:
-            print('{} ({} | {})'.format(white(result),
-                                        blue(result.id),
-                                        blue(format_scores(result))))
+            print('{} ({} | {} | {})'.format(white(result),
+                                             blue(result.id),
+                                             blue(result._hashid),
```
Comment (davidbgk) [inline addok/shell.py]: Is that relevant to expose both?

**Comment by davidbgk** [comment]
I don't get the significative loss in coverage given the few lines of code affected. Thoughts?

**Comment by davidbgk** [comment]
Removing the `_id`  prior to save the document doesn't change significantly the memory consumption: `used_memory_human: 279.08M` (vs. `279.78M` above). Do you think it's worth it for cleaner data @yohanboniface ?

**Comment by davidbgk** [comment]
Using `sqlite-store`:

With hashids:

```
> DBINFO
keyspace_misses: 0
keyspace_hits: 0
used_memory_human: 259.64M
total_commands_processed: 2426074
total_connections_received: 4
connected_clients: 1
nb keys (db 0): 108359
```

Without hashids:

```
> DBINFO
keyspace_misses: 0
keyspace_hits: 0
used_memory_human: 276.62M
total_commands_processed: 7243203
total_connections_received: 13
connected_clients: 1
nb keys (db 0): 108358
```

**Comment by davidbgk** [comment]
To sum up, we have a drop of 6.1% in memory consumption for that dataset, whatever the datastore. It should be even better the more addresses you add.

**Comment by yohanboniface** [comment]
> Removing the _id prior to save the document doesn't change significantly the memory consumption: used_memory_human: 279.08M (vs. 279.78M above). Do you think it's worth it for cleaner data @yohanboniface ?

Nope, let's keep it then

File: addok/ds.py [hunk 5]
```diff
@@ -56,7 +56,9 @@ def store_documents(docs):
     for doc in docs:
         if not doc:
             continue
-        key = keys.document_key(doc['id'])
+        if '_id' not in doc:
```
Comment (yohanboniface) [inline addok/ds.py]: Is it a feature that we don't want to override `_id` if already in the source data? If yes, we may unit test it :)

Comment (davidbgk) [inline addok/ds.py]: Oh, I didn't think about it but I like the idea, might need some doc too!

Comment (davidbgk) [inline addok/ds.py]: @yohanboniface I added https://github.com/addok/addok/pull/307/files#diff-9a94836d2e1beb79dcd5fc085dc5a1a6R94

Comment (davidbgk) [inline addok/ds.py]: @yohanboniface and the documentation https://github.com/addok/addok/pull/307/files#diff-648a1879a1bd05c8f82dcd828b2a9c84R18

File: docs/import.md [hunk 6]
```diff
@@ -15,6 +15,7 @@ The expected keys are the ones declared in the `FIELDS` attribute of your
 - **type** is expected even if not declared as index field
 - **importance** is expected; must be a `float` between 0 and 1
 - **id** is expected
```
Comment (yohanboniface) [inline docs/import.md]: We can remove this line I think.

File: addok/db.py [hunk 7]
```diff
@@ -13,6 +17,10 @@ def connect(self, *args, **kwargs):
     def __getattr__(self, name):
         return getattr(self.instance, name)
 
+    def next_id(self):
+        _counter = self.incr('_counter')
```
Comment (yohanboniface) [inline addok/db.py]: nit: `_id_sequence`?
